Accessibility: Try adding the keyboard navigation/edit mode

youknowriad opened the pull request:
refs #5694

Adding the navigation/edit mode can be very impacting for the edit flow, so it's best to try to do it by steps.

In this PR.

 - tabbing navigate to blocks directly
 - Enter enters "edit" mode 
 - Escape moves back to navigation mode

The drawback of the edit/navigation mode is that it's not possible to create an empty paragraph after the selected block by clicking "Enter", since "Enter" is the key used to move to edit mode. Is this acceptable?

Need help testing the different writing flow interactions to ensure there's no other regression.

youknowriad committed:
Accessibility: Try adding the keyboard navigation/edit mode

youknowriad commented on blocks/library/button/index.js:
```diff
@@ -129,13 +130,15 @@ class ButtonBlock extends Component {
 					onSubmit={ ( event ) => event.preventDefault() }>
 					<Dashicon icon="admin-links" />
 					<UrlInput
+						autoFocus={ false }
```
Without this, the input was being focused when you tab to the button block.

youknowriad commented on components/navigable-container/index.js:
```diff
@@ -65,10 +65,14 @@ class NavigableContainer extends Component {
 
+		if ( this.props.disabled ) {
+			return;
+		}
```
Added a way to disable the tabbableContainer behavior without replacing the component (to avoid rerenders)

youknowriad commented on components/navigable-container/index.js:
```diff
@@ -160,9 +164,9 @@ export class NavigableMenu extends Component {
-			if ( TAB === keyCode ) {
+		const eventToOffset = ( evt, container ) => {
+			const { keyCode, shiftKey, target } = evt;
+			if ( TAB === keyCode && target.parentElement === container ) {
```
The tabbable container behavior triggers only if we're currently focusing one of its children.

youknowriad committed:
Fix unit tests

youknowriad committed:
Fix in-between inserter

mcsf commented on editor/components/block-list/layout.js:
```diff
@@ -217,22 +226,26 @@ class BlockListLayout extends Component {
-						renderBlockMenu={ renderBlockMenu }
-					/>
-				) ) }
+				<TabbableContainer disabled={ this.state.keyboardMode !== 'navigation' } restrictNavigationToChildren>
```
I suspect `restrictNavigationToChildren` is a leftover artifact.

mcsf commented on editor/components/block-list/block.js:
```diff
@@ -173,12 +174,28 @@ export class BlockListBlock extends Component {
+			return;
+		}
+
+		this.focusFirstTextFieldOrContainer( -1 === initialPosition );
```
Minor, but I find readability to be improved by doing:

```js
const isReverse = -1 === initialPosition;
this.focusFirstTextFieldOrContainer( isReverse );
```

jasmussen commented:
🔥  thank you for working on this.

I think the _ingredients_ for something great are here.

The behavior is a little bit confusing, it's easy to lose your place:

![navigation](https://user-images.githubusercontent.com/1204802/37655414-82ca68b2-2c45-11e8-97fe-de091489b2a6.gif)

In this, I start by tabbing through blocks. Enter then _enters_ the block, where pressing the Tab key now cycles the block UI. Only until I press escape, where tab now switches between blocks again. How do I get back to cycling through the block UI? How would I know to press Enter first?

Part of this confusion can be mitigated visually, by not showing the block toolbar or side UI in navigation mode, perhaps even highlighting the block in blue (like multi select) instead of showing the side UI. 

Another part of the confusion is that only tab switches between blocks. Why don't arrowkeys? Given that this is an editor, the arrowkeys are so crucial to moving up and down the content, it feels weird when that jumps straight into edit mode and starts navigating with the caret. It feels brittle and not as modal as I think navigation mode would have to be in order to work.

Windows 10 has an interesting approach to keyboard navigation. **Tab** navigates between _sections_ in a modal. **Arrow keys** then navigates items in that section:

![windows 10](https://user-images.githubusercontent.com/1204802/37655821-9cf3d25e-2c46-11e8-9348-754b5d1cf673.gif)

In the above GIF I'm first tabbing between the three columns of the start menu, then arrow-keying inside each section. 

Perhaps something similar can work for us? 

- You start in navigation mode, and arrowkeys navigate between blocks. Tab and shift tab does the same. 
- Blocks are multi-select blue when selected in navigation mode.
- Enter enters them where you get a caret, and the caret can also be used to navigate between blocks just like it can today.
- In Edit mode, tab is used to access the side UI and block toolbar.

The chief benefit of this approach would be it would _not_ be a new mode we add on top of what we already have. It would essentially be a set of keyboard enhancements to our existing multi select mode. See also this very long document I wrote about that here: https://github.com/WordPress/gutenberg/issues/4382#issuecomment-360108727

Notion.so works similarly to this, if you have questions about how this feels I'd recommend giving that a shot.

The biggest challenge with this branch is that you don't really know that you're in a mode, so the behavior of Tab is unpredictable. It's also confusing because even in edit mode, you can tab through all the blocks, because in this mode, tab tabs through EVERYTHING.

Immediate small steps to try would be:

- In navigation mode, highlight a block same way as a multi selected block. [Like this](https://user-images.githubusercontent.com/1204802/35331014-10d7f63a-0106-11e8-980f-6e9224b81826.png).
- Arrow keys in navigation mode should not enter editing mode, they should switch between blocks, even if Tab and Shift Tab also did this. If you _click text_ or press Enter to enter edit mode, the arrow key would work as it does in master now. 
- When you tab from the Title field to the prompt that currently says "Write your story", you are still in Edit mode, and you can just start typing. 
- Mousing around, keeps you in edit mode.

Future steps to explore could be:

- Should tabbing in edit mode be modal to the block you're editing? You'd still be able to use arrow keys, or press Escape to enter navigation mode, to pick a different block you could then tab through.
- Delete or Backspace when in navigation mode could remove the entire block.
- Escape in navigation mode deselects everything. 
- Multi select with shift also works in navigation mode.

mcsf reviewed (commented):
Nice, this feels like a solid start. I think it's worth keeping this PR small and address the following points later:
- I'm not a screen reader user, and hopefully @afercia will provide better insight, but I have some expectation that pressing Escape when in navigation mode could take me out of the "blocks loop"; perhaps non-intuitively, the only way to move focus back to the WP-Admin sidebar is to tab to the last block, press Enter to switch to edit mode, then tab out of the block list.
- Since pointer actions have no impact on the keyboard mode, behavior may be inconsistent for users who combine pointer and keyboard interactions. For instance, a user clicks into some text, edits, then tabs away; depending on their previous sequence of Enter/Escape input, tabbing may happen in either keyboard mode — seemingly randomly. This is naturally a minor point.

youknowriad committed:
Allow Leaving the navigable container if cycle is false

youknowriad commented:
This is not easy to get right. I agree we should leave this PR small because if we try to make everything right here, this will never land as it's too complex.

So I'm proceeding by small steps and would love if you could tell me what the minimum required to land this.

In my last commit I try to solve this issue:

> I'm not a screen reader user, and hopefully @afercia will provide better insight, but I have some expectation that pressing Escape when in navigation mode could take me out of the "blocks loop"; perhaps non-intuitively, the only way to move focus back to the WP-Admin sidebar is to tab to the last block, press Enter to switch to edit mode, then tab out of the block list.

The approach I took is to avoid cycling through blocks and navigate away when we reach the end. I'm not totally happy with the code though, it seems too complex for a small enhancement. Can't think of an alternative though. What do you think?

jasmussen commented:
Small PRs sound good, and hopefully better devs will chime in with code feedback. Ideally, though, each step along the way brings with it something useful, with as few regressions as possible. Here are some thoughts as to what would be the minimum amount we need necessary in order to progress. 

The chief challenge in this branch as is, is that it's impossible for a user to know when they are in navigation mode, and when they are in edit mode. Since the Tab button acts differently depending on mode, this will cause confusion as to when tab switches between blocks, and when it enters the movers/ellipsis/block toolbar.

In that vein, perhaps in order to move on, the only change we need to do is to hide the mover, the block toolbar, and the ellipsis when in navigation mode. You'd only see the 1px border. 

Everything else could be intact and work as it used to, even exiting navigation mode by using the arrow keys. What do you think?

youknowriad commented:
> In that vein, perhaps in order to move on, the only change we need to do is to hide the mover, the block toolbar, and the ellipsis when in navigation mode. You'd only see the 1px border.

When do you decide to show the movers and the block toolbar in that case since when you use the mouse, to click etc.. the mode is not changed?

jasmussen commented:
> When do you decide to show the movers and the block toolbar in that case since when you use the mouse, to click etc.. the mode is not changed?

As soon as you hover a block, or click a block, you are in edit mode. 

Maybe just moving the mouse puts you in edit mode.

afercia commented:
Thanks for working on this, pretty excited after some first testing.

I'd tend to agree with many of the considerations made here. Trying to proceed step by step, the main thing I see now is the "blocks loop" dilemma. As @mcsf pointed out, there's no way to exit the loop and I totally agree with @youknowriad: there shouldn't be a loop in the first place 🙂 

One of the main goal of this experiment is to put the blocks in a natural tab order. Once on the last block (or the first when shift-tabbing) the expected behavior would be being able to continue tabbing the UI in a natural tab order.

As per the expected interaction when using a screen reader, there's a big difference in terminology and "arrowing" is the default way to explore all content. Without entering into details about [what browse mode and forms mode are](https://tink.uk/understanding-screen-reader-interaction-modes/), I think the best option is trying to emulate the interaction model users would expect in a form, where tabbing navigates through the form controls. I'm not fully convinced arrowing should navigate through blocks, that wouldn't be expected. Comparison with desktop UIs (Windows 10 start menu) is a bit misleading as they're built following an interaction model that is close to the one some ARIA widgets use (e.g. tabs or grids) but it's very different from the one expected in web forms. 

@joen I'd tend to agree in navigation mode the blocks UI shouldn't be displayed. Navigation mode should be just that: navigation :) I'd agree the "modes" should be made clear and communicated to users. For sure, keyboard interaction in a complex application requires documentation and an user guide (see #5420) but I'd tend to agree the modes should be communicated also visually.

> Should tabbing in edit mode be modal to the block you're editing? 
I'd tend to say yes but I guess this could be addressed in a future iteration? See also the initial a11y recommendations in https://github.com/WordPress/gutenberg/issues/297
>  treating each content area as a modal once it has focus might be a good idea. The advantage is that it forces the context to be contained within the editable area.

> The drawback of the edit/navigation mode is that it's not possible to create an empty paragraph after the selected block by clicking "Enter", since "Enter" is the key used to move to edit mode. 

But it is still possible when in Edit mode, right?

Lastly: as a final step, there will be the need to give a role and properly label the container where focus lands in navigation mode. We already use a label for the blocks, I think that should be moved. Currently, when using a screen reader and landing in navigation mode on a block, _all the content_ of the block gets announced because that container has no semantics, it's just a div with a `tabindex="0"`. See below testing with Firefox and NVDA:
- ignore the red border, that's from a plugin I use to test focusable elements
- on the right: all the block content gets announced when focused

![screenshot 86](https://user-images.githubusercontent.com/1682452/37766780-3fceeeb4-2dc8-11e8-85fb-e2b80a0b8e6c.png)

youknowriad commented:
Thanks for your feedback all. 

Before trying to add the visual distinction suggested by @jasmussen between the two modes, I'm trying an alternative technical approach to this issue relying on the WritingFlow component in this branch https://github.com/WordPress/gutenberg/compare/try/writing-flow-edit-mode

I'm thinking it's a better approach that scales better. I'm waiting for feedback here on this second approach before continuing.

jasmussen commented:
Nice work Riad. I tried the try/writing-flow-edit-mode branch, and it seems to work well there as well. 

One thing I noticed in that branch, is that arrowkey navigation doesn't exit edit mode and give you the caret like it does in this branch. I know I noted that wasn't desirable earlier on — but I feel like either the arrow-keys _should_ exit navigation mode and enter edit mode with the caret, _or_ the arrowkeys should work for navigating between blocks in edit mode. 

The more I think about it, though, the more I think (unlike my past assumption) that maybe it's good that the arrow key exits navigation mode. Unsure. But in either case, the arrowkeys should do _something_ in navigation mode. 

Otherwise, seems good!

youknowriad commented:
> One thing I noticed in that branch, is that arrowkey navigation doesn't exit edit mode and give you the caret like it does in this branch.

Good catch and I did it on purpuse :). My idea was that moving the mouse will trigger edit mode, so maybe it's fine that in navigation mode, the arrows do nothing. That said, it's easily tweakable. We can either navigate the blocks like the tabs, keep the caret navigation or do nothing. your choice :).

This flexibility is harder to do in the current branch though :)

jasmussen commented:
Love the sound of that. Then we should go with what seems best _intellectually_ for now, and then tweak and adjust as we test it in master. 👍 👍 🔥

afercia commented:
Nice work! 

Couple things: while we're here, can we please move the `aria-label={ blockLabel }` to the div that gets focused in navigation mode? I'd also add a `role="group"` for now (will check if there's a more appropriate role). I've tested it and good news!
- Safari + VoiceOver
- IE11 + JAWS 2018
- Firefox + JAWS 2018

they all correctly announce the type of the block, which is what we want:

<img width="811" alt="screen shot 2018-03-22 at 14 40 48" src="https://user-images.githubusercontent.com/1682452/37776278-36f7f034-2de5-11e8-97a1-cd56ccf3879c.png">

This is ideal because users will have an immediate feedback while navigating through the blocks. For example:
- press tab -> "Block: Cover Image"
- press tab -> "Block: Paragraph"
- and so on

Instead, Firefox + NVDA 2018.1 do announce, for example, "Block: Paragraph" but they announce also the whole content of the block including the movers, formatting toolbar etc. A bit unfortunate, but seems to me a peculiar behavior of NVDA and can probably be improved using a more appropriate role. Will try to iterate on this.

@youknowriad thanks very much. I've just noticed a small thing, please check if you can reproduce, when you have a chance:
- add a title and a couple paragraphs, for example:
<img width="803" alt="screen shot 2018-03-22 at 14 57 48" src="https://user-images.githubusercontent.com/1682452/37776707-2eda51fc-2de6-11e8-81a4-dd79bbde8694.png">
- in navigation mode, tab to the second paragraph
- tab again
- the "insertion point" button doesn't appear and focus goes to the sidebar (if it's open)

- now Shift+Tab from the sidebar to the blocks list
- the "insertion point" button appears and gets focused
- Shift Tab again: the second paragraph is skipped

Not sure what the best option is, but I'd tend to think the last "insertion point" should always appear.

Of course, the behavior is different if the last block is, for example, an image.

afercia reviewed (commented):
Please see last comment

youknowriad commented:
> - in navigation mode, tab to the second paragraph - tab again - the "insertion point" button doesn't appear and focus goes to the sidebar (if it's open)
now Shift+Tab from the sidebar to the blocks list
the "insertion point" button appears and gets focused
Shift Tab again: the second paragraph is skipped
Not sure what the best option is, but I'd tend to think the last "insertion point" should always appear.

The issue here is that the insertion point is part of the last block (inside the writing flow). What I do know is skip to the end when we reach the last block. But if I don't do this logic, when you reach the last block, you'll go through all its inner tabbable elements before reaching the sidebar (movers, toolbar...)

youknowriad committed:
Try moving the implementation of keyboard/edit mode to the writing flow (#5746)

youknowriad committed:
Remove not needed changes

youknowriad committed:
Move the block's area label

youknowriad committed:
Hiding the block controls on navigation mode and switch to edit mode on mouse move

youknowriad commented:
- I updated the PR based on the WritingFlow component. 
 - Moved the block label to the focusable element + role group
 - Hide all UI aside the border on navigation mode
 - Switch to edit mode as soon as you move the mouse

Let me know what you think about the current state, I'd like to stop enhancements at this level to avoid complexifying more.

afercia commented:
> The issue here is that the insertion point is part of the last block (inside the writing flow).

Ah i Understand. I also realize now the modes are "global", meaning once you switch to edit mode, all blocks are in edit mode. Which is fine, I guess.

Any thoughts about the second paragraph that gets skipped on Shift+Tab?

I understand the need to not introduce further complexity at this point, I'd defer to @jasmussen any decision.

youknowriad committed:
Always select the container when in navigation mode

youknowriad commented:
@afercia Actually, I was able to fix both issues with the last commit. Always skip the inserter in navigation mode and focus the last paragraph.

afercia commented:
Ahh maybe I've catched one thing. Ideally, for keyboard users the default should always be "navigation mode. However, when using Shift+Tab, as soon as I press `Shift` it switches to Edit mode.

youknowriad commented:
> Ahh maybe I've catched one thing. Ideally, for keyboard users the default should always be "navigation mode. However, when using Shift+Tab, as soon as I press Shift it switches to Edit mode.

Actually, it's not the case but since it's hard to start tabbing without moving the mouse first, you end up being in edit mode before maybe?

afercia commented:
You're right, happens only in Safari. As soon as I press Shift, the block UI appears and I'm in edit mode. Can you reproduce?

youknowriad commented:
@afercia I can't seem to reproduce Safari Version 11.0.3

youknowriad commented:
@afercia Wait I do reproduce now. safari 😱

afercia commented:
Yee, see also this short videocapture, where I'm just pressing Shift then Escape, then Shift then Escape... https://cloudup.com/ck5Z_N-JUj8

youknowriad committed:
Fix safari mouse move event

youknowriad commented:
@afercia It should work better now.

afercia commented:
It does! However, seems to me _only on first page load_ it still goes in Edit mode when pressing Shift+Tab.

https://transitory.technology/mouse-move-in-safari/
> The implementation of Safari’s mousemove event differs in a notable way from Chrome and Firefox. After first mouse movement, Safari fires the event on key press of the modifier keys ctrlKey, shiftKey, altKey, and metaKey. 

Can't believe that 🔥

youknowriad commented:
Good catch @afercia it's better now :)

youknowriad commented:
I'd appreciate another review here when y'all have time :)

jasmussen commented:
> I'd appreciate another review here when y'all have time :)

Still testing `try/writing-flow-edit-mode`, let me know if that's the right one to check.

Seems very solid, very nice. 

I still think we need to address, once the technical challenges are addressed, two things:

- in navigation mode, movers, ellipsis, block toolbar should be gone, as an initial way to distinguish between the two modes. 
- arrow keys should always do something — if in navigation mode, they should navigate up and down, or exit to editing mode. No _strong_ opinion here yet, I have thoughts, but they can probably best be formed once we have one or the other.

I did notice that on the very first load, I could navigate the caret but only to the end of the first paragraph. It was like I was in edit mode, and navigation mode, at the same time. Can't quite reproduce, but if I find out how to, I'll note that. 

Still seems like a great approach to me.

youknowriad commented:
@jasmussen Sorry my message was not clear, the branch was merged into this one `try/navigation-edit-mode` and I made the update about the navigation mode UI (arrows not yet).

So better to test `try/navigation-edit-mode` from now on. Thanks

jasmussen commented:
I totally think it works. Very nice, and impressive work! 🔥 

I think before we merge this, we should make the arrow keys do something in navigation mode. Either they should do the same as tab and shift tab, or they should exit navigation mode and move the text caret. Like I mentioned, which makes the most sense is probably best evaluated by simply picking one and trying it. But it does feel a bit off that you have what looks like a selection rectangle in navigation mode, but arrow keys do nothing.

What would be easiest to try?

afercia commented:
Seems to me it works beautifully when using the keyboard. 🎉  I'd tend to second both @jasmussen points: hide the block UI in navigation mode and I'd try to make the arrows navigate up and down.

There's a problem with screen readers on Windows though, same problem being discussed on #5595. 
In short, screen readers on Windows use two "modes": [browse mode and forms mode](https://tink.uk/understanding-screen-reader-interaction-modes/).  In browse mode (the default), they intercept key presses and they pass them back to the browsers only in some circumstances. Basically, the `onKeyDown` used here doesn't fire when using a screen reader.

Will finish testing and comment further (I think it can be solved).

For now, a few things I've noticed while tabbing through blocks in navigation mode. They're probably minor points but the visual effect is a bit confusing:
- post title: the permalink UI appears (note: the permalink UI is still in development and subject to changes)
- cover image block: the cover text formatting toolbar appears
- gallery block: the add item "+" button appears
- embed block: the caption formatting toolbar appears 
- button block: the url input appears

afercia commented:
When in edit mode, and focus is on a button in the formatting toolbar, pressing Escape switches back to navigation mode. This should be avoided as pressing Escape from the formatting toolbar should always move focus back to the block content. I guess stopping propagation would do it?

Instead, when focus is on the movers or the More menu toggle, I guess it's ok to switch back to Edit mode.

afercia commented:
I've tested extensively with screen readers on Windows (building and testing Gutenberg on Windows is a real fun). I'm still undecided about what to do. The keydown event attached to the focusable `div` doesn't work with screen readers in "browse mode" (see previous comment). There are ways to make screen reader stop intercepting key presses: adding some ARIA roles works, for example in the toolbar (role="toolbar") and in the menus (role="menu") the arrow keys work. For the block `div` we could use [role="application"](https://www.w3.org/TR/wai-aria-1.1/#application):

> Some user agents and assistive technologies have a browse mode where standard input events, such as up and down arrow key events, are intercepted and used to control a reading cursor. This browse mode behavior prevents elements that do not have a widget role from receiving and using such keyboard and gesture events to provide interactive functionality.
> When there is a need to create an element with an interaction model that is not supported by any of the WAI-ARIA widget roles, authors MAY give that element role application. And, when a user navigates into an element with role application, assistive technologies that intercept standard input events SHOULD switch to a mode that passes most or all standard input events through to the web application.

But I'm very reluctant to use it because it has potentially devastating effects. For more on this, please refer to this post by Marco Zehe:

**If you use the WAI-ARIA role “application”, please do so wisely!**
https://www.marcozehe.de/2012/02/06/if-you-use-the-wai-aria-role-application-please-do-so-wisely/

I've prepared a quick pen to demo the keydown/click behaviors: it should be tested on Windows for example with Firefox and NVDA: https://codepen.io/afercia/full/NYaWoE/

- without the screen reader running, events fire as expected
- with the screen reader running, keydown doesn't fire; however, `click` does fire.

So I've tried to trigger the switch to "edit mode" also `onClick`, as it seemed to me the simpler way to solve this issue. Unfortunately, for some reason, the `click` event doesn't fire on the Gutenberg block. I'd greatly appreciate some help and feedback on this last point :)

youknowriad committed:
Arrow navigation in navigation mode

youknowriad commented:
Good morning!

I added support for arrow navigation in navigation mode.

> Unfortunately, for some reason, the click event doesn't fire on the Gutenberg block. I'd greatly appreciate some help and feedback on this last point :)

Yes, I just tried and same, I can't figure out why the `Enter` click is not triggering the `onclick` event, maybe it works only on buttons?

> post title: the permalink UI appears (note: the permalink UI is still in development and subject to changes)
> cover image block: the cover text formatting toolbar appears
> gallery block: the add item "+" button appears
> embed block: the caption formatting toolbar appears
> button block: the url input appears

I think we should wait a little bit before doing these changes, These changes mean we need to introduce a new block API (the keyboard mode). This adds some burden on the block author because it means we can't support this consistently across blocks. There's an alternative though: consider the block as not selected even if it's selected in the keyboard mode. Let's see how it goes and update accordingly.

afercia commented:
Here's a quick video of the [Codepen demo](https://codepen.io/afercia/full/NYaWoE/): https://cloudup.com/c40mHOxEmY7

Using the keyboard and pressing Enter on the divs:

Without a screen reader running:
- Enter on the first div: no click and that's expected
- Enter on the second div: keydown fires and that's expected
- the third div is just a variant of the first one

With a screen reader running, things are different, basically the opposite:
- Enter on the first div: fires the click event
- Enter on the second div: no keydown event

afercia commented:
During [today's accessibility meeting on Slack](https://wordpress.slack.com/archives/C02RP4X03/p1522084754000572), the accessibility team has discussed the option to use `role="application"` for the content area of the editor (the list of blocks).

Using `role="application"` would have the immediate effect to switch Windows screen readers in a sort of "forms mode" where they stop intercepting key presses and pass them back to the browser: the `keydown` event used here would work. /cc @youknowriad @aduth 

However, `role="application"` has many (potentially devastating) side effects, as explained above in a previous comment. Please refer to the post by Marco Zehe already mentioned above: worth 10 minutes of your time to get an idea of hat the application role does:

**If you use the WAI-ARIA role “application”, please do so wisely!**
https://www.marcozehe.de/2012/02/06/if-you-use-the-wai-aria-role-application-please-do-so-wisely/

The accessibility team agreed to test this option with A/B testing, involving some of the testers who already tried Gutenberg to check which "mode" provides the best experience for them. I'll take care of preparing a branch to test `role="application"` and report the tests results.

If that doesn't work, there will be the need to find a different solution, as keyboard events don't work as expected with screen readers in "browse mode". I'd encourage to try a different solution anyway. Using a natively interactive control (a button) with a click event would work for everyone.

youknowriad commented:
Can you detail the "button" solution you're proposing? does it mean the block wrapper is a button or you mean adding a button somewhere else to move to edit mode?

Also, how is the application role testing going?

afercia commented:
I meant adding a button somewhere else, but then users should navigate through the buttons, not the wrappers... (I know I know 😱 )

mtias commented:
How close is this to being merged?

youknowriad commented:
I'll defer to @afercia can we use the "application" role?

youknowriad committed:
Fix e2e test (switch to edit mode)

aardrian commented:
> @mtias no decision yet, the "application" role (see #5807) really needs to be tested by assistive technologies users and hopefully some testing will happen at WordCamp London and in the following days. /cc @rianrietveld @aardrian

We have no dedicated AT users on hand at WCLDN for testing. Some testing happened with a skilled screen reader user (and developer) at CSUN (Joe Dolson drove that) and she reported to me separately that the application role was part of her problem. It was eating the keys she relies on to navigate a page, taking an already confusing screen and making it impossible to use.

For non-skilled screen reader users (the only AT users that the application role impacts), I suspect the removal of their SR keys will leave them stranded. This is not based on hard data, just a gut check.

afercia commented:
@aardrian  thanks. Worth noting the version tested at CSUN did _not_ use `role="application"` :) Currently, only the "Classic" block uses `role="application"` because it's basically an instance of TinyMCE which uses it (and it always used it, also in previous WordPress versions).

I do share your concerns though, as many many SR features won't work with `role="application"`.

aardrian commented:
> thanks. Worth noting the version tested at CSUN did not use `role="application"`

Oh dear. Then that makes the feedback I got even more troubling.

afercia commented:
Just to reference the related issue #5694

afercia commented:
For reference and for developers interested in better understanding screen readers browse / forms mode and what `role="application"` does:

Using ARIA
W3C Editor's Draft 22 January 2018
https://w3c.github.io/using-aria/#using-application

afercia commented:
After a quick conversation with @mtias we agreed we should also try to consider alternative approaches. The main goal here is to find a way to greatly simplify keyboard navigation. The real need is to significantly decrease the number of tab stops required to navigate the page blocks. As mentioned in the related issue #5694 say there are 20 blocks in a page:
- with just 20 key strokes users should be able to navigate through all blocks
- when users want to edit a block, there should be a mechanism to enter the block and edit it

So far, the approach considered just using the Tab key, but it could also be a different navigation pattern. Also, the switch between navigation and edit mode is, in a way, an assumption because it's based on a specific workflow.

Other options to evaluate:
- a "global" (at page level) mode switching, based on a keyboard shortcut
- a specific UI control, i.e. a button for each block to switch mode
- anything else that would allow quick navigation through blocks

iamthomasbishop commented:
Hey friends! @mtias asked that I share some of the exploratory (read: not set in stone) design work I've been doing on mobile/native Gutenberg, which features explicit modes to distinguish between "viewing" and "editing" a post/page/doc. These mocks are slightly out-of-date, but the concept looks like this:

(Android)
![gutenberg - general writing flow - example](https://user-images.githubusercontent.com/1200432/40441351-6248987a-5e86-11e8-8474-29c3b3704b36.jpg)

(iOS)
![gutenberg - general writing flow - example](https://user-images.githubusercontent.com/1200432/40441477-c8a65c56-5e86-11e8-9bd5-ff110b65389e.jpg)


The idea originally came about as I was doing a competitive analysis of other mobile editors such as Google Docs, Evernote, and more. I found that splitting up the contexts in this way provides the user more clarity, space, and contextual actions. 

I'm not here to say this would translate 1:1 to the web side, but hopefully it can be of some use as you're exploring options. 😄

afercia commented:
@iamthomasbishop thanks for sharing! It's not so uncommon that mobile and accessibility needs have some points in common, as content is perceived and operated in a linearized way in both cases. 

With regards to keyboard accessibility, there's the need to drastically reduce the number of Tab stops currently necessary to navigate the interface. Navigation/Edit modes (or View/Edit, whatever we want to call them) are an idea to simplify interaction with the blocks, and if that matches the explorations made for mobile, than maybe it's something to explore further.

Question: is the View/Edit mode you are thinking about on a per-block basis or it switches all the blocks in the page at the same time?

iamthomasbishop commented:
> is the View/Edit mode you are thinking about on a per-block basis or it switches all the blocks in the page at the same time?

The idea was that it'd do the latter - switch the entire editor canvas from `view` <=> `edit`. The `view` mode wouldn't show any block controls and would only show actions specific to the task at hand (viewing, so `Publish`, `Post Settings`, etc). From `view` mode, the user would be able to focus a specific block via some mechanism (double-tap or some other accessible alternative) and would thus enter `edit` mode.

My initially feeling was that all of the tabbing/navigating via arrow keys should be done inside of `edit` mode, however now that I think about it, the `view`/`navigation` mode should utilize tabbing to position on a specific block from a “birds-eye view” (in other words, maybe some minimal UI around the highlighted block). From there, focusing the block would activate `edit` mode with that block focused and cursor in it, where you'd be able to use the arrow/equivalent keys to navigate with more granularity.

It's worth noting that I'm neither an accessibility expert nor do I understand the full history/context of this issue, so I hope I'm not way off base here. It also really makes me happy to see such attention to detail around accessibility <3

youknowriad committed:
Fix writing flow after rebase

youknowriad commented:
I rebased this, let's try to get it to the finish line. Even if it doesn't benefit all the users for now, it's still an improvement as is.

> a specific UI control, i.e. a button for each block to switch mode

I think we want to avoid a button per block as it's too cumbersome.

>  a "global" (at page level) mode switching, based on a keyboard shortcut

I'm leaning towards a global keyboard shortcut for now. What would be this shortcut? And do we know if the global shortcuts work on screenreaders?

jasmussen commented:
In the spirit of looking at alternate approaches, here's a very silly question. I expect the answer to be "no, that doesn't make sense", but I'm going to pose the question regardless just in case it inspires something else.

What if you used page down and page up to select blocks. 20 blocks on a page, press page down 20 times to get to the last one.

Everything else is unchanged, you can still tab or use arrow keys.

youknowriad commented:
Or maybe "Ctrl + arrow" instead of "page up/down"?

afercia commented:
> Even if it doesn't benefit all the users for now, it's still an improvement as is.

Hm, hold on please. If the click on the div is the only way to switch to edit mode, this results in the complete inability to edit a block for screen reader users on Windows.

youknowriad commented:
@afercia What's the way forward for you, we have several propositions here and it doesn't make sense to hold this more.

afercia commented:
@youknowriad I've just tested it again on Windows (Firefox + NVDA) and it really creates a huge barrier for screen reader users. When in browse mode, pressing Enter on the div doesn't work. 
Instead, a new block gets added, I guess because the keydown propagates:

![unnamed](https://user-images.githubusercontent.com/1682452/40663941-19bbf1e6-635a-11e8-99fd-b25e0bd03b86.png)

As discussed at length, this is expected with Windows screen readers in browse mode. It's standard behavior and there's nothing to "fix". It's also one of the reasons why jsx-a11y warns to not use events on non-interactive elements.

There are only two ways to make this work I can think of:
- use a proper ARIA role on the div to make screen readers stop intercepting key strokes: unfortunately there are no suitable roles in this case
- use a separate UI control that is natively interactive: a button

A few days ago we've also discussed a bit this issue with @mtias here's some point emerged:
- this button could be the "breadcrumb label" that currently appears only on hover
- alternatively, the switch could be "global", at a page level, i.e. via a global shortcut
- any alternative way to implement the switch is welcome but we can't use the keydown on the div

jasmussen commented:
Let's try and unstick this.

Turns out Slack has a very cool version of this, that I think might work for us. It works like this:

- When the canvas region has focus, up and down arrows _navigate between chat "blocks"_
- When a chat block has focus, you can use tab to navigate through every item for that block, such as star, emoji react, author, more menu, etc.
- You could also tab through everything, but that's a lot of tab stops

Here's a GIF:

![slack keyboard nav](https://user-images.githubusercontent.com/1204802/40827166-1a54e2f0-657d-11e8-93c1-db0f6dd6a38f.gif)

If you noticed, once you started tabbing, a neat little "tip" modal was invoked (cc: @karmatosed) that suggests more keyboard shortcuts:

<img width="550" alt="slack" src="https://user-images.githubusercontent.com/1204802/40827210-3c24dc32-657d-11e8-8fec-c85637454e22.png">

I like this approach because:

- tab works like it would always do, it tabs through everything
- arrow keys work like they do in editors, i.e. they navigate up and down the document, but they double as a block selector
- because of the combination of tabs and arrowkeys, the arrow key navigation can be seen as "additive" on top of the stock tab behavior, and I would suggest it as an intuitive 

I think Andrea's suggestion could make this work for us: 

> use a separate UI control that is natively interactive: a button
> this button could be the "breadcrumb label" that currently appears only on hover

I think we can work with this, especially if #6773 lands first. Imagine if the following GIF showed you using up and down arrows, instead of me mousing over it:

![key nav](https://user-images.githubusercontent.com/1204802/40827556-342f9782-657e-11e8-8882-a707cd6636ee.gif)

Although it looks like the block is selected/hovered, in fact the breadcrumb button is focused. And when you press Enter or space or whatever you need to press to invoke that button, well then you're editing that block. 

CC: @mtias

afercia commented:
Hm... it's an interesting idea. Not sure it will work, as the arrow navigation would need a keydown event on the blocks wrapper div and keyboard events don't work on non-interactive elements with Windows screen readers in browse mode....

jasmussen commented:
Is there any way we can find that out?

afercia commented:
> Is there any way we can find that out?

Well it's basically the same thing that happens in this PR. There would the need of a `keydown` event on the bock wrapper, then filter the key pressed to check for arrows, then fire some function. However, Windows screen readers in browse mode use arrows for their own navigation, they intercept these key presses and the browser would be unaware a keydown occurred.

aardrian commented:
A word of caution — asking screen reader users to switch browse mode on and off is one of the issues that keeps screen reader users from Slack. In short, Slack's interface overall is inaccessible and should be used cautiously for pattern ideas. Related thread from March: https://twitter.com/LeonieWatson/status/974615133044359168

afercia commented:
We've discussed this issue during WCEU contributor day with @aardrian and decided to try a new way: 
- use a visually hidden button with a click event to trigger the mode switch
- the button could be named in a way to convey both the block type information and the associated action, e.g. "Block: Paragraph. Click to edit."
- navigation mode should navigate through these visually hidden buttons, then some CSS should make the block look like "focused"

Not fully sure it is possible from a technical point of view, but it's something definitely worth trying as a native button with a click event would always work even when a screen reader is running.

jasmussen commented:
@afercia @aardrian Per the new way suggested, would we be able to accomplish the proposed direction suggested in https://github.com/WordPress/gutenberg/pull/5709#issuecomment-393791220 ?

Let me know if you'd like me to explain that approach in more detail.

aduth commented:
Is there still a relevancy / desire to continue the work here?

jasmussen commented:
> Is there still a relevancy / desire to continue the work here?

This is still blocked awaiting a response to mockups in https://github.com/WordPress/gutenberg/issues/5694#issuecomment-413798567

tofumatt commented:
Between arrow keys moving through blocks and #10545 giving us a keyboard-accessible (even with Firefox + NVDA) way to move throughout the post's blocks with an easy overview, I think the intent of this PR is covered. I think we should close it for now; please re-open if this is still something we want to look at, but there are issues with this approach that #10545 managed to bypass 🙂
